runtime/src/bank.rs

```diff
@@ -1091,26 +1093,78 @@ impl Bank {
         .unwrap_or_default()
     }
 
-    fn update_clock(&self) {
+    fn update_clock(&self, parent_epoch: Option<Epoch>) {
         let mut unix_timestamp = self.unix_timestamp_from_genesis();
         if self
             .feature_set
             .is_active(&feature_set::timestamp_correction::id())
         {
-            if let Some(timestamp_estimate) = self.get_timestamp_estimate() {
+            let (estimate_type, epoch_start_timestamp) = if self
+                .feature_set
+                .is_active(&feature_set::timestamp_bounding::id())
+            {
+                let timestamp_bounding_activation_slot = self
+                    .get_account(&feature_set::timestamp_bounding::id())
+                    .as_ref()
+                    .and_then(|account| Feature::from_account(account))
+                    .and_then(|feature| feature.activated_at)
+                    .expect("feature has activated");
```

Comment:
Is there an api that does this more elegantly?

Reply:
tendermint + cosmos just use a plain median.  how are you ensuring we have 2/3+ samples for any slot?  there is no guarantee that we have that many votes

Reply:
@aeyakovenko no, we aren't ensuring we have 2/3+ samples for each slot. Do we need to? I think that could be done by extending the `TIMESTAMP_SLOT_RANGE` to include slots back to the root, since we should always have votes from 2/3+ on that slot.

Either way, it still seems like a stake-weighted median is better than a flat median to minimize the possible effect of low-staked validators.

Reply:
could we use all the votes in Tower?  that would ensure that the sample is based on rooted data

Reply:
and to summarize, do we use a weighted 1/3 median?  so drop 1/3 highs and 1/3 lows?

Reply:
> could we use all the votes in Tower? that would ensure that the sample is based on rooted data

We currently don't have access to Tower from the Bank, so that'd require a bit of re-architecting. Not sure I follow why it needs to be rooted data, since the votes will all be in ancestors Banks. Can you tell me more?

As I mentioned above, we can use Bank::ancestors to include all votes/timestamps since the current root, which I think should guarantee 2/3+ stake.

> and to summarize, do we use a weighted 1/3 median? so drop 1/3 highs and 1/3 lows?

I had that written originally, but I updated to just use the stake-weighted median value, so:  the timestamp at which 50% of the stake estimates a greater-or-equal timestamp and 50% of the stake estimates a lesser-or-equal timestamp.

Reply:
@mvines , while this thread has gone off-trail, I did actually want to know if there is a simpler/better way to get an activated feature's activation slot...

Reply:
No better way right now, this is the first case where the activation slot matters.   Can we make it not matter? :)

If no, then I think FeatureSet::active should change from a `HashSet<Pubkey>` to `HashMap<Pubkey,Slot>`, avoid ing all this work on every single new Bank and setting things up for the next time the activation slot matters

Reply:
The ways we could make it not matter involve making assumptions about the realistic range of `epoch_start_timestamp` vs `Bank::get_unused_from_slot`, which I'm not super excited about.

I'll start on that FeatureSet::active edit, if okay with you.

Reply:
sgtm!